Allow to pass bugsnag repo url

stoivo opened the pull request:
This is useful when you want to pass the make a release with source
controll link, with the git relo cloned with ssh but the the gui
expect a http ref

The repo url could be either
	git@github.com:stoivo/bugsnag-capistrano.git
	https://github.com/stoivo/bugsnag-capistrano.git

In my case I use gitlab-onpremise and the params passed to

```
{"apiKey"=>"...",
 "appVersion"=>"...",
 "autoAssignRelease"=>..,
 "builderName"=>"...",
 "buildTool"=>"..",
 "metadata"=>nil,
 "releaseStage"=>"staging",
 "sourceControl"=>
  {"provider"=>"gitlab-onpremise",
   "revision"=>"SHA",
   "repository"=>"https://gitlab.url/org/repo/"}}
```
This might be a bug on the bugsnag server side but this fixes it on the client

## Goal

Make source link work for gitlab on premis wit repo url using git

## Design

Added a config variable

## Changeset

<!-- What changed? -->

## Testing

<!-- How was it tested? What manual and automated tests were
     run/added? -->

yousif-bugsnag commented:
Hi @stoivo,

Can you clarify what issue this PR addresses? Which GUI are you referring to? It should already be possible to use an SSH path for the repo url.

stoivo commented:
I have taken 3 screenshot. The first explains where it is
<img width="1179" alt="example where" src="https://user-images.githubusercontent.com/3492040/116867599-c4b9df00-ac0d-11eb-8695-4e3f98addd57.png">

Second show a bad url when the repo_url are where set to `git@domain:org/repo.git`   
The url is `https://app.bugsnag.com/org/project/git@gitlab.company.no:org/repo.git/blob/a88488815c2c15ac086546d30d0fd3a663dbf172/app/controllers/application_controller.rb#L11` which leads to nothing
![wrong](https://user-images.githubusercontent.com/3492040/116867918-237f5880-ac0e-11eb-979e-c9f4d70617e4.png)



Third is is with this patch and bugsnag_repo_url set to a https://domain/org/repo
![correct](https://user-images.githubusercontent.com/3492040/116868078-6ccfa800-ac0e-11eb-8573-19863d33f6af.png)

xander-jones commented:
hey @stoivo, i've not been able to recreate what you're seeing here. An SSH Git URL is translated as expected for me. Could you please write in to support@bugsnag.com with the `Capfile` config that you were previously using (before this PR) that was causing the trouble? i.e. what were `:repo_url`, `:source_control_provider`, and `:current_revision` set to?

stoivo commented:
I dont think there is anything interestring in Capfile, but there it is

<details>
  <summary>Capfile</summary>

  ```ruby
  # frozen_string_literal: true

  # Load DSL and set up stages
  require "capistrano/setup"

  # Include default deployment tasks
  require "capistrano/deploy"

  # To use Git
  require "capistrano/scm/git"
  install_plugin Capistrano::SCM::Git

  # Include tasks from other gems included in your Gemfile
  #
  # For documentation on these, see for example:
  #
  #   https://github.com/capistrano/rvm
  #   https://github.com/capistrano/rbenv
  #   https://github.com/capistrano/chruby
  #   https://github.com/capistrano/bundler
  #   https://github.com/capistrano/rails
  #   https://github.com/capistrano/passenger
  #
  # require 'capistrano/rails'
  # require "capistrano/rvm"
  require "capistrano/rbenv"
  # require 'capistrano/chruby'
  require "capistrano/bundler"
  require "capistrano/rails/assets"
  require "capistrano/rails/migrations"
  # require 'capistrano/passenger'

  # Load custom tasks from `lib/capistrano/tasks` if you have any defined
  Dir.glob("lib/capistrano/tasks/*.rake").each { |r| import r }

  require "bugsnag-capistrano"
  ```
</details>

On the other hand I think deploy.rb would be interesting. I don't want to share url to our private gitlab instance. So I have replaced the urls with my fork of bugsnag-capistrano

`git@github.com:stoivo/bugsnag-capistrano.git` and `https://github.com/stoivo/bugsnag-capistrano.git`



<details>
  <summary>deploy.rb before </summary>

  ```ruby
  lock "3.14.1"

  set :keep_releases,       100
  set :application,         "application"
  set :user,                "webuser"
  set :ruby_version,        File.read(".ruby-version").strip
  set :deploy_to,           "/var/www/apps/#{fetch :application}"
  set :repo_url,            "git@github.com:stoivo/bugsnag-capistrano.git"

  set :current_local_branch, `git branch`.match(/\* (\S+)\s/m)[1]
  set :current_remote_tracking_branch, `git branch --format="%(HEAD) %(upstream:lstrip=3)"`.match(/\* (\S+)/m).to_a&.pop

  set :rbenv_type, :user # or :system, or :fullstaq (for Fullstaq Ruby), depends on your rbenv setup
  set :rbenv_ruby, File.read(".ruby-version").strip

  set :shared_path, "#{fetch :deploy_to}/shared"
  set :shared_config_path, "#{fetch :shared_path}/config"

  set :default_env, {
    "PATH" => "$PATH:/home/webuser/.nvm/versions/node/v#{File.read(".nvmrc").strip}/bin",
  }

  set :ssh_options, {
    forward_agent: true,
  }

  set :bugsnag_api_key, "bc9c7cfab8c7d4642794a1217903ddb2"
  set :app_version, release_timestamp
  set :bugsnag_auto_assign_release, true

  set :bundle_binstubs, nil
  set :bundle_gemfile,  -> { release_path.join("Gemfile") }
  set :bundle_flags,    ""

  desc "Link config files."
  task :link_config_files do
    on roles(:app, :ws) do
      execute "ln -nfs #{fetch(:shared_config_path)}/database.yml #{fetch(:release_path)}/config/database.yml"
      execute "ln -nfs #{fetch(:shared_config_path)}/secrets.yml #{fetch(:release_path)}/config/secrets.yml"
      execute "ln -nfs #{fetch(:shared_path)}/private #{fetch(:release_path)}/private"
      execute "ln -nfs #{fetch(:shared_path)}/pids #{fetch(:release_path)}/tmp/pids"
      execute "rm -r #{fetch(:release_path)}/log"
      execute "ln -nfs #{fetch(:shared_path)}/log #{fetch(:release_path)}/log"
      execute "ln -nfs #{fetch(:shared_path)}/node_modules #{fetch(:release_path)}/node_modules"
    end
  end

  namespace :deploy do
    task :restart do
      on roles(:app), in: :sequence, wait: 500 do
        execute :touch, release_path.join("tmp/restart.txt")
      end
    end
    task :translations do
      on roles :app do
        within fetch(:release_path) do
          with rails_env: fetch(:rails_env) do
            invoke "bundler:install"
            execute :rake, "i18n:js:export"
          end
        end
      end
    end
  end

  # This section based on
  # https://github.com/mperham/sidekiq/wiki/Deploying-to-Ubuntu#using-capistrano
  # while the sidekiq.service file is based on
  # https://github.com/mperham/sidekiq/blob/master/examples/systemd/sidekiq.service
  namespace :sidekiq do
    task :quiet do
      on roles(:app) do
        # Tells side to stop processing new jobs
        execute "sudo systemctl kill -s TSTP sidekiq-*"
      end
    end
    task :restart do
      on roles(:app) do
        execute "sudo systemctl restart sidekiq-*"
      end
    end
  end
  after "deploy:starting", "sidekiq:quiet"
  after "deploy:reverted", "sidekiq:restart"
  after "deploy:published", "sidekiq:restart"

  after "deploy:updating", "link_config_files"
  after "deploy:updating", "deploy:translations"
  before "deploy:finishing", "deploy:restart"
  ```
</details>


<details>
  <summary>diff to deploy.rb after applying the patch</summary>

```diff
diff --git a/config/deploy.rb b/config/deploy.rb
index 3fce4be95..9881e09a3 100755
--- a/config/deploy.rb
+++ b/config/deploy.rb
@@ -33,6 +33,38 @@ set :bugsnag_api_key, "bc9c7cfab8c7d4642794a1217903ddb2"
 set :app_version, release_timestamp
 set :bugsnag_auto_assign_release, true
 
+set :bugsnag_source_control_provider, "gitlab-onpremise"
+set :bugsnag_repo_url, "https://github.com/stoivo/bugsnag-capistrano/"
+
+namespace :bugsnag do
+  # duplicate  bugsnags release hook to include bugsnag_repo_url
+  # this will trigger the release twice but it don't hurt that much
+  # https://github.com/bugsnag/bugsnag-capistrano/pull/22
+  desc "Notify Bugsnag that new production code has been released"
+  task :release do
+    run_locally do
+      begin
+        auto_assign_release_bool = ENV["BUGSNAG_AUTO_ASSIGN_RELEASE"] == "true"
+        Bugsnag::Capistrano::Release.notify({
+          api_key: fetch(:bugsnag_api_key, ENV["BUGSNAG_API_KEY"]),
+          app_version: fetch(:app_version, ENV["BUGSNAG_APP_VERSION"]),
+          auto_assign_release: fetch(:bugsnag_auto_assign_release, auto_assign_release_bool),
+          builder_name: fetch(:bugsnag_builder, ENV["BUGSNAG_BUILDER_NAME"] || ENV["USER"]),
+          metadata: fetch(:bugsnag_metadata),
+          release_stage: fetch(:bugsnag_env) || ENV["BUGSNAG_RELEASE_STAGE"] || fetch(:rails_env) || fetch(:stage) || "production",
+          revision: fetch(:current_revision, ENV["BUGSNAG_REVISION"]),
+          repository: fetch(:bugsnag_repo_url, fetch(:repo_url, ENV["BUGSNAG_REPOSITORY"])),
+          source_control_provider: fetch(:bugsnag_source_control_provider, ENV["BUGSNAG_SOURCE_CONTROL_PROVIDER"]),
+          endpoint: fetch(:bugsnag_endpoint)
+        })
+        info "Bugsnag release notification complete."
+      rescue
+        error "Bugsnag release notification failed, #{$!.inspect}"
+      end
+    end
+  end
+end
+
 set :bundle_binstubs, nil
 set :bundle_gemfile,  -> { release_path.join("Gemfile") }
 set :bundle_flags,    ""
```
</details>


@xander-jones so we are using gitlab on premise. 

By looking at the json request to the api I think the project url: https://api.bugsnag.com/projects/5aafd03c1b201152bae7b0f6. 
By looking a the json request to the api I think the error id: 608a9b3efae2a90007bc07de. 

One error without the patch id "608aa3b300788c7463c30000" and error_id: 608a9b3efae2a90007bc07de
one error with the patch id: 608af62d0078e680535b0000 and  error_id: 608a9b3efae2a90007bc07de

Should I send an email too?

stoivo commented:
Do you want me to send a email too?

mattdyoung commented:
No, I think this should give us enough info to reproduce the original issue you were seeing and assess the PR. We'll let you know if we have further questions. Thanks!

mattdyoung commented:
@stoivo 
I believe I've been able to reproduce what you were seeing, but I think it should be possible to get this to work without the PR to add `bugsnag_repo_url`.

You can specify the configuration options for bugsnag-capistrano in your `Capfile` or as environment variables:
https://docs.bugsnag.com/build-integrations/capistrano/#configuration-options

Specifying like this in your `Capfile` for an on-prem gitlab instance should work:
```
set :bugsnag_source_control_provider, "gitlab-onpremise"
set :repo_url, "https://gitlab.url/org/repo/"
```
or as env variables:
```
BUGSNAG_SOURCE_CONTROL_PROVIDER=gitlab-onpremise
BUGSNAG_REPOSITORY=https://gitlab.url/org/repo/
```

Could you confirm whether that works in your case? If it does I'll look into improving our documentation for this use case.

stoivo commented:
That would work if we could pull with http, but we cant.

mattdyoung commented:
Ah, I see. So you want `repo_url` to be the `ssh://` url you use to clone but for the Bugsnag UI to display the `https://` url rather than the `ssh://` one. That makes sense - thanks for the PR!

stoivo commented:
I don't want to nag but any updates?, as I have managed to work my way around it should just close the PR?

imjoehaines commented:
Thanks for the PR @stoivo!

I've had to redo our CI setup for this repo as it was no longer working on Travis, so I rebased and merged this in #25

stoivo commented:
Thanks, to you for building bugsnag.
